src/radembedded_async.c

```diff
@@ -0,0 +1,282 @@
+		printf("ERROR: Unable to set authserver.\n");
+		rc_destroy(rh);
+		exit(EXIT_FAILURE);
+	}
+
+	if (rc_add_config(rh, "acctserver", acct_server_ip, "config", 0) != 0) {
+		printf("ERROR: Unable to set acctserver.\n");
+		rc_destroy(rh);
+		exit(EXIT_FAILURE);
+	}
+
+	/* Done setting configuration items */
+
+	/* Read in the dictionary file(s) */
+
+	if (rc_read_dictionary(rh, rc_conf_str(rh, "dictionary")) != 0) {
+		printf("ERROR: Failed to initialize radius dictionary\n");
+		exit(EXIT_FAILURE);
+	}
+	if (rc_avpair_add(rh, &send, PW_ACCT_STATUS_TYPE, &status_type, -1, 0) == NULL) {
+		printf("ERROR: Failed adding Acct-Status-Type: to %d\n", status_type);
+		exit(EXIT_FAILURE);
+	}
+	if (rc_avpair_add(rh, &send, PW_ACCT_SESSION_ID, myuuid, -1, 0) == NULL) {
+		printf("ERROR: Failed adding Acct-Session-ID: to %s\n", myuuid);
+		exit(EXIT_FAILURE);
+	}
+	if (rc_avpair_add(rh, &send, PW_USER_NAME, username, -1, 0) == NULL) {
+		printf("ERROR: Failed adding User-Name: to %s\n", username);
+		exit(EXIT_FAILURE);
+	}
+	if (rc_avpair_add(rh, &send, PW_CALLED_STATION_ID, callto, -1, 0) == NULL) {
+		printf("ERROR: Failed adding Called-Station-ID: to %s\n", callto);
+		exit(EXIT_FAILURE);
+	}
+	if (rc_avpair_add(rh, &send, PW_CALLING_STATION_ID, callfrom, -1, 0) == NULL) {
+		printf("ERROR: Failed adding Calling-Station-ID: to %s\n", callfrom);
+		exit(EXIT_FAILURE);
+	}
+	/* Initialize socket related info in RADIUS Handle */
+	if (rc_apply_config(rh) == -1) {
+		printf("ERROR: Failed to update Radius handle socket info");
+		exit(EXIT_FAILURE);
+	}
+
+
+/* ====================== Sending Asynchronous Requests ===================== */
+
+	/*
+	 * Initialize a new multihandle. Each request will later be enqueued into
+	 * this handle. It is the central component for async-radcli.
+	 */
+	struct rc_multihandle *mhdl = rc_multi_create_multihandle();
+	if (!mhdl) {
+		puts("Could not initialize async-multihandle.");
+		exit(EXIT_FAILURE);
+	}
+
+	/* Send the same request count times */
+	for (i = 0; i < count; i++) {
+		req = rc_async_create_handle_acct(rh, client_port, send);
+		if (!req) {
+			printf("INFO: Creating accounting request %i asynchronously failed."
+				" Terminating.\n", i);
+			exit(EXIT_FAILURE);
+		}
+		if (rc_multi_add(mhdl, req)) {
+			printf("INFO: Enqueing accounting request %i asynchronously failed."
+				" Terminating.\n", i);
+			exit(EXIT_FAILURE);
+		}
+
+		printf("Request 0x%016lx enqueued.\n", (unsigned long)req);
+	}
+
+	/* Kick-off processing of the newly created requests */
+	rc_multi_process(mhdl);
+
+
+	struct pollfd *polli = calloc(count, sizeof(struct pollfd));
+	if (polli == NULL) {
+		puts("ERROR: Failed to allocate pollfds.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	int nr_of_poll_fds = 0;
+	int received_responses = 0;
+
+	for (; received_responses < count;) {
+		/* Get all the currently enqueued pollable file descriptors. */
+		nr_of_poll_fds = rc_multi_get_pollfds(mhdl, polli, count);
```

Comment:
What is my concern is that the API is very poll-system-call-centric. Today common ways to do it is also with epoll, kqueue or libev. How does the API accomodate these?

Reply:
There is also `rc_multi_get_fd_set` and `rc_multi_process_fd_set` to be used with `select`.  

Also, the entire multihandle API is a thin layer on top of the `rc_async_*` API to provide an easy way to handle multiple radius requests with poll or select.  Those syscalls always report the entire set and multihandle helps with walking and processing the pollfds array or select fdset. It's loosely inspired by libcurl but much simpler.

`rc_async_*` can also directly be used with and event multiplexer syscall as it exposes the associated fd and the expected events in any state.

Extending the multihandle API to more complex event multiplexer syscalls like epoll and kqueue might be possible although I'm not sure how useful that would be given those syscalls report events not for the entire set of descriptors but only for descriptors that saw events.

Reply:
Indeed, we should not provide wrappers for all possibilities, just make sure that it is possible to use the API with them. Documentation seem to be important towards that, so that the async API is not perceived as poll-only.